File: unity/Assets/Scripts/Multiplayer/Rooms/Actions/PendingAction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;
using static Michitai.Multiplayer.Time.Time;

namespace Michitai.Multiplayer.Rooms.Actions
{
    /// <summary>
    /// Information about a pending action that needs to be completed by the host.
    /// Contains action details, sender information, and request data.
    /// Uses Unity's JsonUtility for deserialization of request data.
    /// </summary>
    /// <typeparam name="T">The type of request data.</typeparam>
    [System.Serializable]
    public class PendingAction<T> where T : class, new()
    {
        [SerializeField]
        private string request_data_json;    // Unity mode
        [SerializeField]
        private string created_at;

        /// <summary>
        /// The unique ID of the action.
        /// </summary>
        public string action_id;

        /// <summary>
        /// The ID of the player who submitted the action.
        /// </summary>
        public int player_id;

        /// <summary>
        /// The ID of the target player.
        /// </summary>
        public int target_id;

        /// <summary>
        /// The type of action.
        /// </summary>
        public string action_type;

        /// <summary>
        /// The name of the player who submitted the action.
        /// </summary>
        public string player_name;

        /// <summary>
        /// Whether the action was submitted by the host.
        /// </summary>
        public bool is_host;

        /// <summary>
        /// The request data deserialized into the specified type.
        /// </summary>
        public T RequestData
        {
            get
            {
                return JsonUtility.FromJson<T>(request_data_json);
            }
        }

        /// <summary>
        /// Timestamp when the action was created.
        /// </summary>
        public DateTimeOffset? CreatedAt
        {
            get
            {
                return ParseUtc(created_at);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/actions/PendingAction.java
package com.michitai.multiplayer.rooms.actions;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Information about a pending action awaiting completion.
 *
 * @param <T> The type to deserialize action request data into.
 */
public class PendingAction<T> {
    @JsonProperty("action_id")
    private String actionId;

    @JsonProperty("player_id")
    private int playerId;

    @JsonProperty("target_id")
    private int targetId;

    @JsonProperty("player_name")
    private String playerName;

    @JsonProperty("is_host")
    private boolean isHost;

    @JsonProperty("action_type")
    private String actionType;

    @JsonProperty("request_data_json")
    private String requestDataJson;

    @JsonProperty("request_data")
    private T requestData;

    @JsonProperty("created_at")
    private String createdAt;

    public String getActionId() {
        return actionId;
    }

    public void setActionId(String actionId) {
        this.actionId = actionId;
    }

    public int getPlayerId() {
        return playerId;
    }

    public void setPlayerId(int playerId) {
        this.playerId = playerId;
    }

    public int getTargetId() {
        return targetId;
    }

    public void setTargetId(int targetId) {
        this.targetId = targetId;
    }

    public String getPlayerName() {
        return playerName;
    }

    public void setPlayerName(String playerName) {
        this.playerName = playerName;
    }

    public boolean isHost() {
        return isHost;
    }

    public void setHost(boolean host) {
        isHost = host;
    }

    public String getActionType() {
        return actionType;
    }

    public void setActionType(String actionType) {
        this.actionType = actionType;
    }

    public String getRequestDataJson() {
        return requestDataJson;
    }

    public void setRequestDataJson(String requestDataJson) {
        this.requestDataJson = requestDataJson;
    }

    public T getRequestData() {
        return requestData;
    }

    public void setRequestData(T requestData) {
        this.requestData = requestData;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }
}
